blackjax/inference/base.py

```diff
@@ -0,0 +1,200 @@
+"""Base kernel for the HMC family."""
+from typing import Callable, Dict, List, NamedTuple, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+
+from blackjax.inference.proposals import HMCProposalInfo, HMCProposalState
+
+__all__ = ["HMCState", "HMCInfo", "hmc"]
+
+PyTree = Union[Dict, List, Tuple]
+
+
+class HMCState(NamedTuple):
+    """State of the HMC algorithm.
+
+    The HMC algorithm takes one position of the chain and returns another
+    position. In order to make computations more efficient, we also store
+    the current potential energy as well as the current gradient of the
+    potential energy.
+    """
+
+    position: PyTree
+    potential_energy: float
+    potential_energy_grad: PyTree
+
+
+class HMCInfo(NamedTuple):
+    """Additional information on the HMC transition.
+
+    This additional information can be used for debugging or computing
+    diagnostics.
+
+    acceptance_probability
+        The acceptance probability of the transition, linked to the energy
+        difference between the original and the proposed states.
+    is_accepted
+        Whether the proposed position was accepted or the original position
+        was returned.
+    is_divergent
+        Whether the difference in energy between the original and the new state
+        exceeded the divergence threshold.
+    energy
+        The total energy that corresponds to the returned state.
+    proposal
+        The state proposed by the proposal. Typically includes the position and
+        momentum.
+    proposal_info
+        Information returned by the proposal. Typically includes the step size,
+        number of integration steps and intermediate states.
+    """
+
+    acceptance_probability: float
+    is_accepted: bool
+    is_divergent: bool
+    energy: float
+    proposal: HMCProposalState
+    proposal_info: HMCProposalInfo
+
+
+def hmc(
+    potential_fn: Callable,
+    proposal_generator: Callable,
```

Comment:
I don't know how far you want to go with type hints, but perhaps they could be added here as well? e.g.
```
    proposal_generator: Callable[[Array, HMCProposalState], Tuple[HMCProposalState, HMCProposalInfo]],
```
(if I've understood correctly)

(I can work on adding these if you're open to them)

Reply:
I think it could eventually be useful to users to have their IDE raise a warning if the function does not have the correct signature. Let's wait  until NUTS to see if the signature stays the same :)

Reply:
OK, sure!